### PR TITLE
chore: cache qrcode

### PR DESCRIPTION
**Pull request by zhouhanseng**

### CODE REVIEW

File: src/components/shared/qrcode.tsx [hunk 1]
```diff
@@ -32,8 +32,9 @@ export function QRCode({ text, options = {}, canvasProps }: QRProps) {
         if (!ref.current) return
 
         qr.toCanvas(ref.current, text, options, () => {
```
Comment (septs) [inline src/components/shared/qrcode.tsx:34]: https://www.npmjs.com/package/qrcode#tocanvascanvaselement-text-options-cberror

```suggestion
        qr.toCanvas(ref.current, text, options, (error) => {
            if (error) { .... }
```

**Review by septs — CHANGES_REQUESTED** [review]

File: src/components/shared/qrcode.tsx (hunk 1 above)
Comment (septs) [reply src/components/shared/qrcode.tsx:34]: or https://www.npmjs.com/package/qrcode#todataurltext-options-cberror-url ?

Comment (zhouhanseng) [reply src/components/shared/qrcode.tsx]: done

File: src/components/shared/qrcode.tsx [hunk 2]
```diff
@@ -31,9 +31,11 @@ export function QRCode({ text, options = {}, canvasProps }: QRProps) {
     useEffect(() => {
         if (!ref.current) return
 
-        qr.toCanvas(ref.current, text, options, () => {
-            if (ref.current?.toDataURL()) {
-                cache.set(text, ref.current?.toDataURL())
+        qr.toCanvas(ref.current, text, options, (err: Error) => {
+            if (err) throw err
```
Comment (septs) [inline src/components/shared/qrcode.tsx:35]: Replace QR Code area show a error report,
e.q: QRCode Render failed. (add to i18n)

**Review by septs — CHANGES_REQUESTED** [review]

File: src/components/shared/qrcode.tsx (hunk 2 above)
Comment (septs) [reply src/components/shared/qrcode.tsx:35]: CC @neruthes need design a copywriting.
1. `This function is temporarily unavailable`
2. `QRCode Render failed`
do you have a better idea?

Comment (neruthes) [reply src/components/shared/qrcode.tsx:35]: I think that "Unknown error; please try again." will be good.

Comment (septs) [reply src/components/shared/qrcode.tsx:35]: OK, The function current is silent failure is a bad UX.
i think he can solve it.

Comment (septs) [reply src/components/shared/qrcode.tsx:35]: > I think that "Unknown error; please try again." will be good.

Click the text, try re-render the qrcode?

Comment (neruthes) [reply src/components/shared/qrcode.tsx:35]: > Click the text, try re-render the qrcode?

Making the "try again" part be clickable with certain styling (color & text-decoration underline) will be good. The user can click "try again" in the error message text line to try again.

File: src/components/shared/qrcode.tsx [hunk 3]
```diff
@@ -31,9 +31,11 @@ export function QRCode({ text, options = {}, canvasProps }: QRProps) {
...
+            const dataURL = ref.current?.toDataURL()
+            if (dataURL) {
+                cache.set(text, dataURL)
             }
         })
     }, [options, text])
```
Comment (Jack-Works) [inline src/components/shared/qrcode.tsx:41]: options👀

Comment (zhouhanseng) [reply src/components/shared/qrcode.tsx:41]: You mean should the options also be cached?

File: src/_locales/en/messages.json [hunk 4]
```diff
@@ -99,6 +99,7 @@
     "error_gas_feed_exceeds": "Error occurred when invoking the smart contract.",
     "error_insufficient_balance": "Your allowance in this wallet is not sufficient to do that.",
     "error_unknown": "Unknown Error.",
+    "error_unknown_please": "Unknown Error; please ",
```
Comment (septs) [inline src/_locales/en/messages.json:102]: see https://react.i18next.com/latest/trans-component

```suggestion
    "error_unknown_please": "Unknown Error; please <again>try again</again>",
```

reference code:

https://github.com/DimensionDev/Maskbook/blob/c1cc8d28933337817490206cbcd8fa3f21846cdf/src/plugins/Gitcoin/DonateDialog.tsx#L176-L181

**Review by septs — CHANGES_REQUESTED** [review]

File: src/components/shared/qrcode.tsx [hunk 5]
```diff
@@ -13,31 +18,69 @@ interface QRProps {
...
+        if (!ref.current || error) return
 
-        qr.toCanvas(ref.current, text, options, () => {
-            if (ref.current?.toDataURL()) {
-                cache.set(text, ref.current?.toDataURL())
+        qr.toCanvas(ref.current, text, options, (err: Error) => {
+            if (err) {
+                setError(true)
+                cache.remove(text)
+                console.error(err)
```
Comment (septs) [inline src/components/shared/qrcode.tsx:57]: plz remove debugging code.

File: src/components/shared/qrcode.tsx [hunk 6]
```diff
@@ -13,31 +18,69 @@ interface QRProps {
 }
 
 const cache = new Proxy(sessionStorage, {
```
Comment (septs) [inline src/components/shared/qrcode.tsx:20]:
```typescript
const cache = {
   get() { ... },
   set() { ... },
   remove() { ... }
}
```

**Review by septs — CHANGES_REQUESTED** [review]

File: src/components/shared/qrcode.tsx (hunk 6 above)
Comment (zhouhanseng) [reply src/components/shared/qrcode.tsx]: done

File: src/components/shared/qrcode.tsx (hunk 5 above)
Comment (zhouhanseng) [reply src/components/shared/qrcode.tsx]: done

File: src/_locales/en/messages.json (hunk 4 above)
Comment (zhouhanseng) [reply src/_locales/en/messages.json]: done

File: src/_locales/en/messages.json [hunk 7]
```diff
@@ -35,6 +35,7 @@
     "backup_persona": "Backup Your Persona",
     "backup_wallet": "Backup Wallet",
     "backup_wallet_hint": "Keep the <strong>12 mnemonic words</strong> OR <strong>private key</strong> below <strong>carefully</strong> in a safe place. You will need them to restore this wallet.",
+    "backup_qrcode_error": "Unknown Error; please <TryAgain>try again</TryAgain>.",
```
Comment (septs) [inline src/_locales/en/messages.json:38]:
```suggestion
    "backup_qrcode_error": "Unknown Error; please <again>try again</again>.",
```

**Review by septs — CHANGES_REQUESTED** [review]

File: src/components/shared/qrcode.tsx [hunk 8]
```diff
@@ -12,32 +18,60 @@ interface QRProps {
...
-    get(t, p: 'get' | 'set') {
-        return {
-            get(key: string) {
-                return t.getItem(`qrcode:${key}`)
-            },
-            set(key: string, value: string) {
-                return t.setItem(`qrcode:${key}`, value)
-            },
-        }[p]
+const cacheScope = 'qrcode'
```
Comment (septs) [inline src/components/shared/qrcode.tsx:21]:
```suggestion
const CACHE_SCOPE = 'qrcode'
```

**Review by septs — CHANGES_REQUESTED** [review]

File: src/components/shared/qrcode.tsx (hunk 8 above)
Comment (zhouhanseng) [reply src/components/shared/qrcode.tsx]: done

File: src/_locales/en/messages.json (hunk 7 above)
Comment (zhouhanseng) [reply src/_locales/en/messages.json]: done

**Review by septs — APPROVED** [review]

**Comment by septs** [comment]
@neruthes @Jack-Works re-review.

**Review by Jack-Works — APPROVED** [review]

**Pull request merged**